### PR TITLE
feat(UIForms): add collapsiblePanel

### PR DESCRIPTION
**Pull request by jsomsanith**
**What is the problem this PR is trying to solve?**
_Warning: this is only on UIForms, not current forms implementation_
Missing Collapsible Fieldset

**What is the chosen solution to this problem?**
Add it, in a standalone mode, and integrate it with arrays

**Please check if the PR fulfills these requirements**
- [x] The PR commit message follows our [guidelines](https://github.com/talend/tools/blob/master/tools-root-github/CONTRIBUTING.md)
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Related design / discussions / pages (not in jira), if any, are all linked or available in the PR

**[ ] This PR introduces a breaking change**

### CODE REVIEW

File: output/forms.eslint.txt [hunk 1]
```diff
@@ -13,5 +13,8 @@ The react/require-extension rule is deprecated. Please use the import/extensions
 /home/travis/build/Talend/ui/packages/forms/src/UIForm/fields/Radios/Radios.component.js
   23:7  error  Form controls using a label to identify them must be programmatically associated with the control using htmlFor  jsx-a11y/label-has-for
 
-✖ 3 problems (3 errors, 0 warnings)
+/home/travis/build/Talend/ui/packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js
+  48:7  error  Visible, non-interactive elements should not have mouse or keyboard event listeners  jsx-a11y/no-static-element-interactions
```
Comment (jmfrancois) [inline output/forms.eslint.txt]: I guess you should escape this one

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			event.stopPropagation();
+			event.preventDefault();
+			const payload = {
+				schema: this.props.schema,
+				value: {
+					...this.props.value,
+					isClosed: !this.props.value.isClosed,
+				},
+			};
+			this.props.onChange(event, payload);
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: the onChange props is required but has no default value
is it injected by the Form ?

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+		}
+
+		render() {
+			const { id, schema, value, ...restProps } = this.props;
+			const { items } = schema;
+			const iconTransform = !this.props.value.isClosed ? 'flip-vertical' : '';
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: `classNames({'flip-vertical': !this.props.value.isClosed})`

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+			return (
+				<fieldset
+					className={classNames('form-group', theme['collapsible-panel'], 'collapsible-panel')}
+				>
+					<div
+						onDoubleClick={this.toggle}
+						id={id && `${id}__title_bar`}
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: if no id I have id={false} :/

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+						role="button"
+						className={theme['title-bar']}
+					>
+						<div
+							onClick={this.toggle}
+							id={id && `${id}__title_wrapper`}
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: same here

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+							role="button"
+							className={theme.title}
+						>
+							<legend id={id && `${id}__title`}>{title(value, schema)}</legend>
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: same here

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+						</div>
+						<button
+							onClick={this.toggle}
+							id={id && `${id}__collapse`}
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: same here

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.scss [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+		background: $white;
+		box-shadow: 0 7px 14px rgba(0, 0, 0, 0.25), 0 5px 5px rgba(0, 0, 0, 0.22);
+		border: 1px solid $gallery;
+	}
+
+	.title-bar {
+		display: flex;
+		align-items: stretch;
+		justify-content: space-between;
+		border-bottom: 1px solid #f6f6f6;
```
Comment (jmfrancois) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.scss]: $wild-sand plesae

File: packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js [hunk 9]
```diff
@@ -31,36 +21,40 @@ export default function ArrayItem(props) {
...
+										previousIndex: index,
+										nextIndex: index + 1,
+									})
+								}
+								type="button"
+								title="Move Down"
+							>
+								<Icon name="talend-caret-down" />
+							</button>
+						)}
```
Comment (talend-glorieux) [inline packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js]: That code is duplicated. Please refactor.

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js [hunk 10]
```diff
@@ -0,0 +1,94 @@
+import PropTypes from 'prop-types';
+import React from 'react';
+import Icon from '@talend/react-components/lib/Icon';
+import classNames from 'classnames';
+import Widget from '../../Widget';
+
+import theme from './CollapsibleFieldset.scss';
+
+function defaultTitle(_, schema) {
```
Comment (talend-glorieux) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: Why the underscore?

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+	cases(
+		'should toggle',
+		opts => {
+			// given
+			const CollapsibleFieldset = createCollapsibleFieldset();
+			const onChange = jest.fn();
+			const event = {
+				stopPropagation: jest.fn(),
+				preventDefault: jest.fn(),
+			};
```
Comment (talend-glorieux) [inline packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js]: Can't you extract all the options cases independent constant outside of the function block?

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js (hunk 2 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: Yes it is injected by the forms. It's the callback that will change/consolidate the model

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js (hunk 4 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: if no id, then `id={undefined}`

File: output/forms.eslint.txt (hunk 1 above)
Comment (frassinier) [reply output/forms.eslint.txt]: no.

File: packages/forms/src/UIForm/fieldsets/Array/Array.component.js [hunk 12]
```diff
@@ -126,7 +120,7 @@ export default class ArrayWidget extends React.Component {
 		const canReorder = schema.reorder !== false;
 
 		return (
-			<div>
+			<div className={theme['tf-array-container']}>
```
Comment (frassinier) [inline packages/forms/src/UIForm/fieldsets/Array/Array.component.js]: may you double className as plain text?

File: packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js (hunk 9 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js]: Not really, those are the 2 reorder arrows. Refactoring it to create a common piece will imply to be able to customize the disabled, id, onClick, title, icon classname.
Basically, all the props except 2 : the `type="button"` and the icon name.

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js (hunk 10 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: It's to say `whatever the name is, I don't use it in this implem`

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js (hunk 11 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js]: They are only constants within the function block. It allows to not have interference between tests.
In the tests, I check if the function has been called. If not initialized on each test (so not constants), the assertion will be always true after the first test.

File: packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js (hunk 9 above)
Comment (talend-glorieux) [reply packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js]: Can't you just pass the an `up` or `down` value that generates the proper disabled, id, nextIndex, title and icon?

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js (hunk 10 above)
Comment (talend-glorieux) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: Then why is the argument here?

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js (hunk 11 above)
Comment (talend-glorieux) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.test.js]: mOK

File: packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js (hunk 10 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: It's the value.
* the value is the 1st arg for retrocompat
* In arrays, each value is the item, so you can create a title depending on each item (you need it in your project)

Comment (talend-glorieux) [reply packages/forms/src/UIForm/fieldsets/CollapsibleFieldset/CollapsibleFieldset.component.js]: OK. I just hope it doesn't become a trend :wink:

File: packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js (hunk 9 above)
Comment (jsomsanith) [reply packages/forms/src/UIForm/fieldsets/Array/ArrayItem.component.js]: done

**Pull request merged**